cmd/gpu-kubelet-plugin/device_health.go

```diff
@@ -0,0 +1,342 @@
+	if ret := m.eventSet.Free(); ret != nvml.SUCCESS {
+		klog.Warningf("failed to unset events: %v", ret)
+	}
+
+	if ret := m.nvmllib.Shutdown(); ret != nvml.SUCCESS {
+		klog.Warningf("failed to shutdown NVML: %v", ret)
+	}
+	close(m.unhealthy)
+}
+
+func (m *nvmlDeviceHealthMonitor) run(ctx context.Context) {
+	defer m.wg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			klog.V(6).Info("Stopping event-driven GPU health monitor...")
+			return
+		default:
+			event, ret := m.eventSet.Wait(5000) // timeout in 5000 ms.
+			if ret == nvml.ERROR_TIMEOUT {
+				continue
+			}
+			// not all return errors are handled as currently there is no proper way to process these errors other than marking all devices healthy.
+			// Ref doc: [https://docs.nvidia.com/deploy/nvml-api/group__nvmlEvents.html#group__nvmlEvents_1g9714b0ca9a34c7a7780f87fee16b205c].
+			if ret != nvml.SUCCESS {
+				if ret == nvml.ERROR_GPU_IS_LOST {
+					klog.Infof("GPU is lost error: %v; Marking all devices as unhealthy", ret)
+					m.markAllDevicesUnhealthy()
+					continue
+				}
+				klog.Warningf("Error waiting for NVML event: %v. Retrying...", ret)
+				continue
+			}
+
+			// TODO: check why other supported types are not considered?
+			eType := event.EventType
+			eData := event.EventData
+			eGI := event.GpuInstanceId
+			eCI := event.ComputeInstanceId
+			if eType != nvml.EventTypeXidCriticalError {
+				klog.Infof("Skipping non-nvmlEventTypeXidCriticalError event: Data=%d, Type=%d, GI=%d, CI=%d", eData, eType, eGI, eCI)
+				continue
+			}
+
+			if m.skippedXids[eData] {
+				klog.Infof("Skipping XID event: Data=%d, Type=%d, GI=%d, CI=%d", eData, eType, eGI, eCI)
+				continue
+			}
+
+			klog.Infof("Processing event XID=%d event", eData)
+			// this seems an extreme action.
+			// should we just log the error and proceed anyway.
+			// TODO: look into how to properly handle this error.
+			eventUUID, ret := event.Device.GetUUID()
+			if ret != nvml.SUCCESS {
+				klog.Infof("Failed to determine uuid for event %v: %v; Marking all devices as unhealthy.", event, ret)
+				m.markAllDevicesUnhealthy()
+				continue
+			}
+			affectedDevice := m.deviceByPlacement.get(eventUUID, eGI, eCI)
+			if affectedDevice == nil {
+				klog.Infof("Ignoring event for unexpected device (UUID:%s, GI:%d, CI:%d)", eventUUID, eGI, eCI)
+				continue
+			}
+
+			klog.Infof("Sending unhealthy notification for device %s due to event type:%v and event data:%d", affectedDevice.UUID(), eType, eData)
+			m.unhealthy <- affectedDevice
+		}
+	}
+}
+
+func (m *nvmlDeviceHealthMonitor) Unhealthy() <-chan *AllocatableDevice {
+	return m.unhealthy
+}
+
+func (m *nvmlDeviceHealthMonitor) markAllDevicesUnhealthy() {
+	for _, giMap := range m.deviceByPlacement {
+		m.markAllSlicesUnhealthy(giMap)
+	}
+}
+
+// markAllSlicesUnhealthy is a helper function to mark every mig device under a parent as unhealthy.
+func (m *nvmlDeviceHealthMonitor) markAllSlicesUnhealthy(giMap map[uint32]map[uint32]*AllocatableDevice) {
```

Comment:
The docstring is clear, but the method name is not clear.

Let's rename to something like `markAllMigDevicesUnhealthy`.

"slice" in the context of the DRA driver in my mind refers to a resource slice.

Reply:
done